ext/drreg/drreg.c

```diff
@@ -65,14 +65,17 @@
 
 #define PRE instrlist_meta_preinsert
 
-/* The number of spills should be pretty hard to exceed as there aren't
- * that many GPRs. We have a hard-coded max on spill slots that we track.
- */
-#define MAX_SPILLS (SPILL_SLOT_MAX + 8)
-/* We choose the number of available slots for spilling simds to match their
- * theoretical max number for a given build.
+/* We have a hard-coded max on spill slots that we track.
+ * The number of spills should be pretty hard to exceed as there aren't
+ * that many GPRs.
+ *
+ * Note that additional slots are required to act as temporary storage
```

Comment:
I actually think we added the 8 slots for in case the client already uses the available spill slots, so there are 8 available for drreg. Also it sounds misleading, because the client has to actively request additional spill slots for cross-app, this just makes them available. Also the comment about the simd slots was lost?

Reply:
I don't understand the main concern apart from the removal of the SIMD comment. I agree with the following statement:

>  actually think we added the 8 slots for in case the client already uses the available spill slots, so there are 8 available for drreg.

The client uses all slots indicated by SPILL_SLOT_MAX, e.g. it reserves all GPR registers. The client then needs additional storage, e.g. when handling cross-app values, in order to function. What is invalidating the comment exactly?

I will re-add the SIMD comment. I removed it as I coalesced the code and felt it was not needed.

Reply:
I think I should rephrase that additional slots *can* be used for temporary storage as an example, because the comment comes across like they are there strictly for that purpose atm. Is this what you mean?

Reply:
My main doubt is that SPILL_SLOT_MAX is the client max for DR's slots. DR's slots are not guaranteed cross-app. So if drreg has to resort to DR slots, then cross-app spilling like this is broken already. The comment says "and we document that tools should request an extra slot for each cross-app-instr register", so the tool is responsible to allocate additional raw tls space. Are you saying you think this makes sure the tool-requested slots are trackable? I don't see why they wouldn't be if the 8 was missing, as long as the tool had requested enough raw tls slots.

Unrelated: should we add DR_REG_START_GPR - DR_REG_STOP_GPR + 1 number of registers instead of 8?

Reply:
> I think I should rephrase that additional slots can be used for temporary storage as an example, because the comment comes across like they are there strictly for that purpose atm. Is this what you mean?

That, too. But also, I think the number of trackable slots here is just a hardcoded number with no strong hard limit to the additional slots drreg needs for cross-app. There is no strong relation to SPILL_SLOT_MAX, because drreg allocates raw tls which is used primarily. Maybe re-phrase and say exactly that, plus mention that drreg needs to be able to track additional slots cross-app. Maybe add a XXX comment mentioning that cross-app spilling has an issue once raw tls slots are filled up.

Reply:
Okay, I understand the issue and will be providing a fix.

Reply:
> My main doubt is that SPILL_SLOT_MAX is the client max for DR's slots. DR's slots are not guaranteed cross-app. So if drreg has to resort to DR slots, then cross-app spilling like this is broken already. 

Yes, for GPRs; using DR slots will break things. This is not the case for SIMD however.

Reply:
One question, perhaps one can point me to the code in DynamoRIO, but is it safe for drreg to actually exceed SPILL_SLOT_MAX?

Reply:
Done. Let me know if it needs fixing.

Reply:
> One question, perhaps one can point me to the code in DynamoRIO, but is it safe for drreg to actually exceed SPILL_SLOT_MAX?

Any idea on this please? Also, I still am not confident about this hard limit in general. How do we know that +8 is enough? At least there are some checks to ensure we do not go beyond the limit in the current code-base.

Reply:
As long as the tool reserves enough ops.num_spill_slots, which it should, I don't see a problem.

Reply:
Yes, this is just an upper bound to simplify tracking data structures.  The logic used to compute it uses existing spill slot bounds in other interfaces just to try to arrive at some reasonable upper bound.  It is not meant to have implications for DR-vs-raw slots or cross-app behavior.

Reply:
Okay 👌 Thanks for clearing that up. Essentially, the user MUST reserve enough ops.num_spill_slots. My concern was that the user does not, for instance allocate any raw slots and relies solely on DR slots (assuming no cross app spilling), and therefore runs the risk of exceeding SPILL_SLOT_MAX due to many register reservations. However, this concern is addressed if the user does indeed specify enough nim_spill_slots.